sdks/java/io/rrio/src/test/java/org/apache/beam/io/requestresponse/EchoHTTPCallerTestIT.java

```diff
@@ -0,0 +1,113 @@
+import static org.junit.Assert.assertThrows;
+import static org.junit.Assert.assertTrue;
+
+import com.google.protobuf.ByteString;
+import java.net.URI;
+import org.apache.beam.testinfra.mockapis.echo.v1.Echo.EchoRequest;
+import org.apache.beam.testinfra.mockapis.echo.v1.Echo.EchoResponse;
+import org.apache.beam.testinfra.mockapis.echo.v1.EchoServiceGrpc;
+import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Tests for {@link EchoHTTPCallerTestIT} on a deployed {@link EchoServiceGrpc} instance's HTTP
+ * handler. See {@link EchoITOptions} for details on the required parameters and how to provide
+ * these for running integration tests.
+ */
+@RunWith(JUnit4.class)
+public class EchoHTTPCallerTestIT {
+
+  private static @MonotonicNonNull EchoITOptions options;
+  private static @MonotonicNonNull EchoHTTPCaller client;
+  private static final ByteString PAYLOAD = ByteString.copyFromUtf8("payload");
+
+  @BeforeClass
+  public static void setUp() throws UserCodeExecutionException {
+    options = readIOTestPipelineOptions(EchoITOptions.class);
+    if (options.getHttpEndpointAddress().isEmpty()) {
+      throw new RuntimeException(
+          "--httpEndpointAddress is missing. See " + EchoITOptions.class + "for details.");
+    }
+    client = EchoHTTPCaller.of(URI.create(options.getHttpEndpointAddress()));
+
+    EchoRequest request = createShouldExceedQuotaRequest();
+
+    // Quota allocation needs at least 1 so we invoke a few times to guarantee exceeding quota.
+    // We only throw exceptions that are not UserCodeQuotaException.
+    try {
+      EchoResponse ignored = client.call(request);
+      client.call(request);
+      client.call(request);
+      client.call(request);
+    } catch (UserCodeExecutionException ignored) {
+    }
+  }
+
+  @Test
+  public void givenValidRequest_receivesResponse() throws UserCodeExecutionException {
+    EchoRequest request = createShouldNeverExceedQuotaRequest();
+    EchoResponse response = client.call(request);
+    assertEquals(response.getId(), request.getId());
+    assertEquals(response.getPayload(), request.getPayload());
+  }
+
+  @Test
+  public void givenExceededQuota_shouldThrow() {
+    assertThrows(UserCodeQuotaException.class, () -> client.call(createShouldExceedQuotaRequest()));
+  }
+
+  @Test
+  public void givenNotFound_shouldThrow() {
+    UserCodeExecutionException error =
+        assertThrows(
+            UserCodeExecutionException.class,
+            () ->
+                client.call(
+                    EchoRequest.newBuilder()
+                        .setId("i-dont-exist-quota-id")
+                        .setPayload(PAYLOAD)
+                        .build()));
+
+    assertTrue(error.getMessage().contains("404 Not Found"));
```

Comment:
Just curious if we have/can use something like Mockito?

We could use more idiomatic tests, for example 
```
assertThat(error.getMessage(), containsString("404 Not Found"));
```

It looks like the test is the same, but failure reports would be (just an example):

> String is 429 Too Many Requests, expected to contain 404 Not Found

Instead of:

> Expected true is false

Reply:
> Just curious if we have/can use something like Mockito?
There are mocking tests in a separate PR that are validating catching specific errors and mock the Caller and SetupTeardown interfaces for this purpose. Tests in this PR validate the integration of the Caller and SetupTeardown implementation and a real API built and deployed in a cloud environment.

Reply:
> We could use more idiomatic tests

I agree. I will change the PR to follow this.

Reply:
Sorry, I meant Hamcrest instead of Mockito.

Reply:
I agree. Hamcrest seems much more readable. I will change over to using that. Thank you, Bruno.